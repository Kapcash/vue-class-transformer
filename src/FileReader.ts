import fs from 'fs';
import ts from 'typescript';
import VueCompiler, { SFCDescriptor } from 'vue-template-compiler'


function isVueFile(filePath: string): boolean {
  return filePath.endsWith('.vue')
} 

/** Recursively get all files from the given folder
 * If the given path is a file, it just returns the file
 * @param path The file or folder path
 * @return All the files present in the folder and its sub-folder, or the file if it's not a folder
 */
function getFilesFromFolder (path = "./") {
  const entries = fs.readdirSync(path, { withFileTypes: true });

  const vueFiles = entries
    .filter(file => !file.isDirectory())
    .map(file => `${path}/${file.name}`)
    .filter(isVueFile);
  const folders = entries.filter(folder => folder.isDirectory());

  for (const folder of folders) {
    vueFiles.push(...getFilesFromFolder(`${path}/${folder.name}`));
  }
  return vueFiles;
}

/** Get all files paths from path and sub-folders */
function getAllFilesPaths (fileOrFolderPath): string[] {
  if (!fs.existsSync(fileOrFolderPath)) throw new Error('Invalid path provided')
  
  if (fs.lstatSync(fileOrFolderPath).isDirectory()) {
    return getFilesFromFolder(fileOrFolderPath)
  } else {
    if (!isVueFile(fileOrFolderPath)) throw new Error('This is not a .vue file!')
    return [fileOrFolderPath]
  };
}

export function getAllFilesToUpgrade (path: string): FileDescriptor[] {
  return getAllFilesPaths(path)
    .map(path => new FileDescriptor(path))
    .filter(file => !!file)
}

export function getSfcDescriptor (file: FileDescriptor): SFCDescriptor | null {
  let sourceStr = fs.readFileSync(file.fullPath, 'utf8')
  
  if (file.isVueFile) {
    return VueCompiler.parseComponent(sourceStr)
  } else {
    return null
  }
}

export function extractScriptFromSfc(sfc: SFCDescriptor): ts.SourceFile {
  return ts.createSourceFile('inline.ts', sfc.script.content, ts.ScriptTarget.ES2020)
}

/** A class descripting a file main attributes */
export class FileDescriptor {
  /** @see https://regex101.com/r/skRAGV/1 */
  private MATCH_SCRIPT_FILE_PATH = /^(.*\/)?(.+)\.(vue|ts|js)$/

  fullPath: string = ''
  path: string = ''
  name: string = ''
  extension: string = ''
  
  constructor(fullPath: string) {
    const [ _fullPath, path, name, extension ] = fullPath.match(this.MATCH_SCRIPT_FILE_PATH)

    if (!_fullPath) return null

    this.fullPath = fullPath
    this.path = path
    this.name = name
    this.extension = extension
  }

  get isVueFile(): boolean {
    return this.extension === 'vue'
  }

  get nameWithExtension(): string {
    return `${this.name}.${this.extension}`
  }
}
