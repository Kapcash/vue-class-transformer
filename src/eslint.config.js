module.exports = {
  "space-before-blocks": "warn",
  "padding-line-between-statements": [
    "error",
    { blankLine: "always", prev: "*", next: "*"},
  ],
  "@typescript-eslint/member-delimiter-style": ["error"],
  semi: "off",
  "@typescript-eslint/semi": ["error", "always"],
  "no-return-assign": ["error", "except-parens"],
  "comma-dangle": ["error", "always-multiline"],
  "lines-between-class-members": ["error", "always", { exceptAfterSingleLine: true }],
  "quote-props": ["error", "as-needed"],
  "space-before-function-paren": ["error", {
    anonymous: "never",
    named: "never",
    asyncArrow: "always",
  }],
  "import/no-mutable-exports": "off",
  "no-unused-vars": "off",
  "import/order": "off",
  "no-console": "off",
  camelcase: "off",
}
