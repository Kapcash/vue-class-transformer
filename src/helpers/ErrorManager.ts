export class ErrorManager {
  errorMap = new Map<string, Error[]>();

  addError(fileName: string, message: Error) {
    if (this.errorMap.has(fileName)) {
      const errorEntry = this.errorMap.get(fileName);
      errorEntry.push(message);
    } else {
      this.errorMap.set(fileName, [message]);
    }
  }

  toString() {
    const errorLogs: string[] = [];
    for(const [name, errors] of this.errorMap.entries()) {
      errorLogs.push(`- ${name}: ${errors.map(err => err.stack || err.message)}`);
    }
    return errorLogs.join('\n\n');
  }
}