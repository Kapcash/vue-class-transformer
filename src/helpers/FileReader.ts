import fs from 'fs';
import { dirname } from "path";
import mkdirp from 'mkdirp'
import ts from 'typescript';
import * as VueCompiler from 'vue-template-compiler'
import { SFCDescriptor, SFCBlock } from 'vue-template-compiler'

/** Recursively get all files from the given folder
 * If the given path is a file, it just returns the file
 * @param path The file or folder path
 * @return All the files present in the folder and its sub-folder, or the file if it's not a folder
 */
function getFilesFromFolder (path = ".") {
  const entries = fs.readdirSync(path, { withFileTypes: true });

  const folderFiles = entries
    .filter(file => !file.isDirectory())
    .map(file => `${path}/${file.name}`)
  const folders = entries.filter(folder => folder.isDirectory());

  for (const folder of folders) {
    folderFiles.push(...getFilesFromFolder(`${path}/${folder.name}`));
  }
  return folderFiles;
}

/** Get all files paths from path and sub-folders */
function getAllFilesPaths (fileOrFolderPath): string[] {
  if (!fs.existsSync(fileOrFolderPath)) throw new Error('Invalid path provided')
  
  if (fs.lstatSync(fileOrFolderPath).isDirectory()) {
    return getFilesFromFolder(fileOrFolderPath)
  } else {
    return [fileOrFolderPath]
  };
}

export function getAllFilesToUpgrade (path: string): FileDescriptor[] {
  path = `${process.cwd()}/${path}`
  const vueFiles = getAllFilesPaths(path)
    .map(path => new FileDescriptor(path))
    .filter(file => file.isVueFile)
  if (vueFiles.length <= 0) { throw new Error("This path doesn't include any .vue file!"); }
  return vueFiles;
}

export function getSfcDescriptor (file: FileDescriptor): SFCDescriptor | null {
  let sourceStr = fs.readFileSync(file.fullPath, 'utf8')
  
  if (file.isVueFile) {
    return VueCompiler.parseComponent(sourceStr)
  } else {
    return null
  }
}

export function extractScriptFromSfc(sfc: SFCDescriptor): ts.SourceFile {
  return ts.createSourceFile('inline.ts', sfc.script.content, ts.ScriptTarget.ES2020)
}

export function createFileAndFolders(filePath: string, content: string | NodeJS.ArrayBufferView) {
  mkdirp.sync(dirname(filePath))
  fs.writeFileSync(filePath, content)
}

export function replaceVueScript(filePath: string, outputPath: string, scriptBlock: SFCBlock, scriptString: string) {
  mkdirp.sync(dirname(filePath))
  fs.open(filePath, 'r', function(err, fd) {
    if (err) { throw 'could not open file: ' + err; }

    const stats = fs.fstatSync(fd);

    const beforeSize = scriptBlock.start
    const beforeBuffer = Buffer.alloc(beforeSize)
    fs.readSync(fd, beforeBuffer, null, beforeSize, null)

    const afterSize = stats.size - scriptBlock.end
    const afterBuffer = Buffer.alloc(afterSize)
    fs.readSync(fd, afterBuffer, null, afterSize, scriptBlock.end)

    const scriptBuffer = Buffer.from(scriptString, 'utf8')

    fs.writeFileSync(outputPath, Buffer.concat([beforeBuffer, scriptBuffer, afterBuffer]), { encoding: 'utf-8' })
  })
}

/** A class descripting a file main attributes */
export class FileDescriptor {
  /** @see https://regex101.com/r/skRAGV/1 */
  private MATCH_SCRIPT_FILE_PATH = /^(.*\/)?(.+)\.(vue|ts|js)$/

  fullPath: string = ''
  path: string = ''
  name: string = ''
  extension: string = ''
  
  constructor(fullPath: string) {
    const [ _fullPath, path, name, extension ] = fullPath.match(this.MATCH_SCRIPT_FILE_PATH)

    if (!_fullPath) return null

    this.fullPath = fullPath
    this.path = path
    this.name = name
    this.extension = extension
  }

  get isVueFile(): boolean {
    return this.extension === 'vue'
  }

  get nameWithExtension(): string {
    return `${this.name}.${this.extension}`
  }
}
