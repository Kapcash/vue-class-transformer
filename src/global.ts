import { RuntimeConfiguration } from './helpers/ArgumentParser';
import { ErrorManager } from './helpers/ErrorManager';

export type SFCToken = 'script' | 'template' | 'styles' | 'other';
export type AttributeToken = 'data' | 'props' | 'computed' | 'watcher' | 'hooks' | 'methods' | 'other';

export function Alias(parameter: AttributeToken | SFCToken) {
  return function (target, propertyKey: string, descriptor: PropertyDescriptor) {
    target[parameter] = descriptor.value;
  };
}

Array.prototype.chunk = function(statement) {
  return this.reduce(
    ([validChunk, invalidChunk], value) => {
      const targetChunk = statement(value) ? validChunk : invalidChunk;
      targetChunk.push(value);
      return [validChunk, invalidChunk];
    },
    [[], []]
  );
};

declare global {
  namespace NodeJS {
    interface Global {
      config: RuntimeConfiguration;
      errors: ErrorManager;
    }
  }

  interface Array<T> {
    chunk(statement: (val: T) => boolean): [Array<T>, Array<T>];
  }
}
