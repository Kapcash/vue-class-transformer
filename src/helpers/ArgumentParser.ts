import minimist from 'minimist'
import chalk from 'chalk';
import { AttributeToken, SFCToken } from '../global.js';

export interface Configuration {
  overrideFiles: boolean;
  isNuxt: boolean;
  outputDir: string;
  sfcOrder: SFCToken[];
  propertiesOrder: AttributeToken[];
}

export interface RuntimeConfiguration extends Configuration {
  inputPath: string;
}

export async function parseArguments(args): Promise<RuntimeConfiguration> {
  // Arguments parsing
  const argv = minimist(args, {
    boolean: ['n', 'f', 'h'],
    alias: {
      n: 'nuxt',
      h: 'help',
      f: 'force',
      c: 'config',
      o: 'output',
    },
    default: {
      n: false,
      f: false,
      o: './generated',
    }
  })

  const { _: argPath, config: configPath, help, nuxt, force, output } = argv
  
  if (help) {
    return Promise.reject(false)
  }
    
  let options: Configuration = {
    isNuxt: nuxt,
    overrideFiles: force,
    outputDir: output,
    sfcOrder: ['script', 'template', 'styles', 'other'],
    propertiesOrder: ['data', 'props', 'watcher', 'hooks', 'methods', 'computed', 'other'],
  }

  if (configPath) {
    const configAbsPath = `${process.cwd()}/${configPath}`
    try {
      const configFile: Configuration = (await import(configAbsPath)).default
      options = { ...options, ...configFile }
    } catch (err) {
      console.error(`Can't find the configuration file at ${configAbsPath}.`, err)
    }
  }
  
  const inputPath = argPath[0]
  const outputDir = options.outputDir.endsWith('/') ? options.outputDir : `${options.outputDir}/`

  const runtimeConfig: RuntimeConfiguration = {
    ...options,
    outputDir,
    inputPath,
  }
  
  return runtimeConfig;
}

export function printHelp() {
  const info = console.log;
  info(`${chalk.bold('Syntax:')}   vts [options] [file|folder path]`)
  
  info('')

  info(`Example:  vts -n -f ./src/test.vue`)
  info(`Example:  vts -o dist/ ./src/components/`)
  
  info('')

  info(chalk.bold(`Options:`))
  info(` -n, --nuxt          Import decorators from 'nuxt-property-decorator' library. Default is 'vue-property-decorator'.`)
  info(` -f, --force         Replace converted files in place instead of created new files.`)
  info(` -o, --output        Path to the output folder. Ignored if option --force is used.`)
  info(` -c, --config         Path to a configuration file. The configuration file options override CLI options if they collide.`)
  info(` -h, --help          Print this helper.`)
  info(` -v, --version       Print the CLI current version.`)
  
  info('')

  info(`See the github README.md for more information: https://github.com/Kapcash/vue-typescript-converter#readme`)
}