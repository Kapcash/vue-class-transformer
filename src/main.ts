import ts from "typescript";
import { SFCDescriptor } from "vue-template-compiler";
import { sourceExtractor } from "./extractors/NodeExtractor";
import VueComponentDescriptor from "./extractors/VueComponentDescriptor";
import { Configuration } from "./helpers/ArgumentParser";
import { FileDescriptor, getSfcDescriptor, replaceVueScript, extractScriptFromSfc } from "./helpers/FileReader";
import VuePropertyDecoratorBuilder from "./builders/component/ComponentBuilders";
import ComponentDirector from "./builders/component/ComponentDirector";

/** Transform a Vue SFC to another syntax, defined by a director */
export function upgradeComponent(config: Configuration) {
  return function (vueFile: FileDescriptor) {
    const sfc = getSfcDescriptor(vueFile)
    const sourceScript = convertScript(config, sfc)

    const outputPath = config.overrideFiles ? vueFile.fullPath : `${config.outputDir}${vueFile.nameWithExtension}`

    replaceVueScript(vueFile.fullPath, outputPath, sfc.script, sourceScript)
  }
}

/** Convert the given SFC descriptor's script */
export function convertScript(config: Configuration, sfc: SFCDescriptor): string {
  const vueDescriptor = new VueComponentDescriptor()

  const tsSource = extractScriptFromSfc(sfc)

  // For every first level child in the source, we extract all the info we need
  ts.forEachChild(tsSource, sourceExtractor(vueDescriptor))
  
  const componentBuilder = new VuePropertyDecoratorBuilder(config.isNuxt)
  componentBuilder.setDescriptor(vueDescriptor)
  const director = new ComponentDirector(componentBuilder, config.propertiesOrder)
  const newScript = director.build()

  return ts.createPrinter().printFile(newScript)
}