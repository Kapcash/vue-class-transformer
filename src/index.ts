import { getAllFilesToUpgrade, FileDescriptor } from './helpers/FileReader';
import { parseArguments, printHelp, RuntimeConfiguration } from './helpers/ArgumentParser';
import { upgradeComponent } from './main'

// ===== MAIN ===== //

(async function() {
  let config: RuntimeConfiguration
  try {
    config = await parseArguments(process.argv.slice(2))
  } catch (err) {
    if (err === false) {
      printHelp()
      process.exit(0)
    } else {
      console.error("Failed to parse the input arguments.", err)
      process.exit(9)
    }
  }
  const vueFilesToUpgrade: FileDescriptor[] = getAllFilesToUpgrade(config.inputPath)
  vueFilesToUpgrade.forEach(upgradeComponent(config))
})()

// ===================== //