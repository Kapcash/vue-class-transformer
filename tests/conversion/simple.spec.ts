import { Configuration } from '../../src/helpers/ArgumentParser';
import { convertScript } from '../../src/main';
import { getSfcDescriptor, FileDescriptor } from '../../src/helpers/FileReader';

describe('auth', () => {
  const testPath = process.cwd() + '/tests/test.vue'
  const testConfig: Configuration = {
    outputDir: '',
    overrideFiles: false,
    isNuxt: false,
    sfcOrder: ['script', 'template', 'styles', 'other'],
    propertiesOrder: ['data', 'props', 'watcher', 'hooks', 'methods', 'computed', 'other'],
  }

  it('should resolve with true and valid userId for hardcoded token', () => {
    const sfc = getSfcDescriptor(new FileDescriptor(testPath))
    const convertedScript = convertScript(testConfig, sfc)
    expect(convertedScript).toMatchSnapshot();
  })
})