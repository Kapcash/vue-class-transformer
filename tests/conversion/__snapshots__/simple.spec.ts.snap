// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`auth should resolve with true and valid userId for hardcoded token 1`] = `
"import { Vue, Component, Watch, Prop } from 'vue-property-decorator';
import { PasswordField, Test } from \\"~/components/login/fields/password-field.vue\\";
import Tooltip from \\"~/components/login/tooltip.vue\\";
@Component<ResetPasswordModal>({ components: { PasswordField, Tooltip } })
export default class ResetPasswordModal extends Vue {
    isOpen: boolean = this.$store.state.modals.isResetPasswordActive;
    newPassword: string = \\"\\";
    confirmedPassword: string = \\"\\";
    isFormValid: boolean = false;
    showServerTooltip: boolean = false;
    @Prop({
        type: Boolean, required: false, default: false
    })
    readonly loading!: boolean;
    @Prop({
        type: Object, required: true
    })
    readonly val!: Object;
    @Watch('$route')
    on$routeUpdate(newVal, oldVal: string) {
        this.foo = newVal;
    }
    @Watch('foo')
    onFooUpdate(newVal, oldVal: string) {
        this.foo = newVal;
    }
    async updatePassword(newPassword: string): Promise<void> {
        this.showServerTooltip = false;
        try {
            await this.$vrc().accounts().updateAccount({ password: newPassword });
            this.$store.commit(\\"user/SET_PROFILE_PROPERTY\\", { property: \\"password\\", newValue: newPassword });
            this.window = 1;
        }
        catch (error) {
            this.showServerTooltip = true;
            console.error(error);
        }
    }
    close(): void {
        this.$store.commit(\\"modals/SET_RESET_PASSWORD_ACTIVE\\", false);
        this.$router.replace({ path: \\"/\\" });
    }
    get arePasswordsEquals(): boolean {
        return this.newPassword === this.confirmedPassword;
    }
    set arePasswordsEquals(val) {
        return val;
    }
    get arePasswordsFilled(): boolean {
        return this.newPassword !== \\"\\" && this.confirmedPassword !== \\"\\";
    }
    created() {
        this.watcher = this.$store.watch(() => this.$store.state.modals.isResetPasswordActive, (newVal: boolean) => {
            this.isOpen = newVal;
        });
    }
    head() {
        return { meta: \\"test\\" };
    }
}
"
`;
